### DISCUSSION TITLE
Query re moving from CLI to Pythonic usage of streamlink?

### DISCUSSION THREAD

**Discussion by danj2k** [streamlink/streamlink#6334] [Q&A] [answered]
### Description

Because #5601 is currently languishing in pull request limbo, at the moment if I want to record a 24-hour DASH live stream in 30-minute segments, I have to use a batch file similar to this, combining `streamlink` with `ffmpeg`:

```
@echo off
echo.
:top
streamlink --stream-segment-attempts 1 --stream-segment-timeout 10 --stream-segment-threads 10 "https://any-24-hour-live-stream-I-want-to-record-in-30-second-segments.com/stream.mpd" best -O | ffmpeg -i pipe:0 -t 00:30:00 -c copy "Name of Stream (%date:~-4,4%-%date:~-7,2%-%date:~-10,2% - %time:~0,2%_%time:~3,2%_%time:~6,2%).mp4"
echo.
goto top
```

Recently it occurred to me that both streamlink and ffmpeg have Python library versions available, so is there a way I could do the same kind of thing as the above but in Python?

I tried looking at the documentation but I don't really get how I could do it without having an improperly terminated video file every time.

**Comment by bastimeyer** [comment]
The output of [`DASHStream.open()`](https://streamlink.github.io/api/stream.html#streamlink.stream.Stream.open) is still a `StreamIO(io.BaseIO)` instance, which means that while `.read()`ing it, you can't `.close()` it at specific segments, because the segments of each sub-stream were already concatenated and all sub-streams were already muxed into the single output stream.

**Reply by danj2k** [reply]
OK, but is there a Pythonic equivalent of piping that into ffmpeg, similar to how the above CLI example works? And have ffmpeg worry about finding a frame boundary at the end of each of my desired segments?

**Reply by bastimeyer** [reply, marked as answer]
- https://docs.python.org/3/library/subprocess.html#subprocess.Popen
- https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stdin

Example:  
https://github.com/streamlink/streamlink/blob/7.0.0/src/streamlink_cli/streamrunner.py

FFmpeg should discard garbage data at the end of the (stopped) stream output.

You'll have to figure all of this out on your own though. I can't help you with your own implementation. Good luck.
